### PR TITLE
testing/mcp: read declared properties back through debug info

### LINKED ISSUE #13251: MCP: declared property values can't be read back, so verifying state means diffing pixels
Split out from #13244, which reported four verification gaps together.

## Problem

An MCP client can drive a UI but can't read back what a declared property holds.
`get_element_properties` returns `absolutePosition`, `size`, `accessibleLabel`, `accessibleRole`, `computedOpacity` and `typeNamesAndIds`.
`set_element_value` writes the *accessible* value only.
`ElementHandle`'s accessors in `internal/backends/testing/search_api.rs` are accessibility plus structure, so the MCP layer isn't declining to expose something it already has.

Verifying that an interaction produced the intended state therefore means screenshotting before and after and diffing pixels.
That answers "these pixels changed", not "this property holds this value", and it says nothing about state with no distinct visual.
In the case that prompted #13244 each verification step moved ~350 KB of base64 PNG, and the diff only became reliable after zooming the UI so the change cleared a noise threshold.

## Why this isn't an MCP-layer fix

Name-based property access exists only in the interpreter: `ComponentInstance::get_property` (`internal/interpreter/component.rs:236`, `internal/interpreter/api.rs:1512`).
For a compiled Rust or C++ application — what the embedded MCP server usually runs in — `ItemTreeVTable` exposes `accessible_string_property` and no generic property access.
There is no table mapping a property name to its storage in generated code, so there is nothing for the introspection layer to read.

Two consequences for the design:

1. Supporting compiled applications means generating a property table, in the spirit of what `SLINT_EMIT_DEBUG_INFO` already emits for type names and ids. That is a codegen decision with a binary-size cost, and it needs an opt-in story.
2. Even under the interpreter, `get_property` reaches the root component's exposed properties. The case in #13244 was a `selected` flag on a repeated row, which is a property of a sub-component instance, so it isn't reachable by that path either.

## Possible directions

- Emit a name-to-property table under the existing debug-info flag, and expose typed reads through `ElementHandle` and a `properties` field on `get_element_properties`.
- Restrict the feature to interpreter-hosted applications, such as `slint-viewer --features mcp`, and say so plainly in the tool description.
- Decline it, and document the accessible-value surface plus `start_event_recording` as the supported verification path.

Value encoding is a real question in any of these, though `set_element_value` and `accessible_value` already establish a convention to extend.

## Related

- #13244 — the original report of this and three other gaps.

### PR DESCRIPTION
**Pull request by tilladam**
An MCP client (and any `ElementHandle` user) can now read an element's declared `in`/`out`/`in-out` properties back, so verifying UI state no longer means diffing screenshots.

Closes #13251. Addresses gap 1 of #13244.

## Design

- Under the existing debug-info flag (`SLINT_EMIT_DEBUG_INFO=1` / `with_debug_info`), the compiler records each element's readable declared properties at LLR lowering (`SubComponent::element_properties`).
  `move_declarations` captures the source name and root key of every declaration it hoists, ranking the effective shadowing declaration first; the LLR optimizer filters and remaps the table like every other reference holder.
- Two new `ItemTreeVTable` entries expose the table, following the `accessible_string_property` stringify convention: `element_declared_properties` enumerates `name:type` pairs, `element_property_value` reads one value in the string encoding of the new `i_slint_core::debug_info` module.
  The contract is tri-state: `false` means unsupported, `true` with an empty result means nothing declared, so a runtime without the feature is distinguishable from an element without properties.
- All three runtimes serve the same table with the same encoding: generated Rust bakes match arms mirroring `item_element_infos`, the interpreter evaluates through `eval::load_property`, and generated C++ dispatches through `slint::private_api` helpers.
  C++ delegates float encoding to the Rust runtime over FFI: GCC 10 (a supported toolchain) has no floating-point `std::to_chars`, and no local formatter reproduces Rust's `f32` `Display` digits for values like `1e20`.
- `ElementHandle` gains `declared_properties()` and `declared_property_value()` in the Rust and C++ testing APIs.
  `get_element_properties` and `get_element_tree` report `declaredProperties` with typed JSON values, and say so through a note when debug info is missing instead of degrading silently.

Example response fragment:

```json
"declaredProperties": [
  { "name": "selected", "typeName": "bool", "value": true },
  { "name": "len", "typeName": "length", "value": 12 },
  { "name": "tint", "typeName": "color", "value": "#112233ff" }
]
```

Properties the compiler optimized out (constant, or never read) are not listed, and private declarations are excluded, unless marked `@testable` (below). Writing declared properties is left to a follow-up.

## The `@testable` attribute

Without it, a test depends on the optimizer not removing the property, so any new optimization would be a breaking change for tests (tronical's point in the thread). `@testable` turns that into a contract:

- It joins `@deprecated` and `@shadowable` in the member-attribute slot, gated experimental like `@shadowable`. Only properties accept it: callbacks, functions, global and interface properties are rejected with a diagnostic.
- A `@testable` property is listed regardless of visibility (a `private` one included) and survives both removal stages: `remove_unused_properties` keeps it when debug info is on, and in the LLR its `ElementProperty` is pinned so `count_property_use` counts it as a real read. Language visibility is unchanged.
- It guarantees listing and a live binding, not a readable value for every type: a `@testable` struct property lists with no value, like any other struct.
- Without debug info, the property is removed like any other unused one, so release builds are unaffected (pinned by a unit test).
- Docs: `guide/experimental/testable`.

## Testing

- One source, one set of expected strings, run on all three runtimes: `tests/cases/testing/declared_properties.slint` (generated Rust and C++) and `internal/interpreter/tests/declared_properties.rs` (interpreter), covering every encodable type, float edge values (`0.1`, `1e20`), aliases, inherited and use-site declarations, repeated rows, reactivity, and the exclusions.
- Shadow resolution (`@shadowable`) is pinned by a compiler unit test, as are the four `@testable` invariants: listed regardless of visibility, binding kept, removed without debug info, no second entry for a shadowed base declaration.
- The shared case gains a private unread `@testable` int (listed with its value) and a private unread `@testable` struct (listed, no value), run on all three runtimes.
- Full runs: interpreter driver (891), Rust driver (808), C++ driver (862), syntax tests, and the testing backend's MCP tests, all green.

## Cost

Measured on a release build of the gallery (macOS arm64):

| | no flag | with `SLINT_EMIT_DEBUG_INFO=1` |
|---|---|---|
| master | 19,282,672 B | 19,374,720 B |
| this branch | 19,372,304 B | 19,598,960 B |

Flag off, the two vtable entries and their stubs cost +89,632 B (+0.46%); flag on, the tables and value-read code cost +224,240 B (+1.16%).

### CODE REVIEW

**Comment by tronical** [comment]
> Properties the compiler optimized out (constant, or never read) are not listed; private declarations are excluded.

In the past, this has been indeed the primary concern about this feature. If a new Slint version introduces a new optimisation that optimises a property away that was previously accessible and now isn't anymore, then that would constitute a breaking change - preventing us from making the optimisation. That's why we were thinking of introducing an extra attribute to a property to indicate that this particular property is required for the purpose of system testing. So not necessarily part of the component's public API, but relevant to testing.

> Writing declared properties is left to a follow-up.

Can you elaborate on the follow-up plans?

**Comment by tilladam** [comment]
Right, the PR leans on that limitation being acceptable for now, and the docs' "anchor the property with a `changed` callback" advice is exactly that: a workaround, not a contract. The attribute you describe is the missing contract, and I think it layers cleanly on top of this PR:

- the attribute's only job is to keep the property alive through the optimizer: exempt it from const-propagation and unused-removal when debug info is on, the way a `changed` callback already does implicitly;
- everything downstream is unchanged: the lowering-time table, the vtable contract, and the encoding don't change shape. A pinned property just reliably appears in the table it already flows through.

So tests written against pinned properties would get a stability guarantee instead of today's "hope it survives the optimizer", and the introspection channel doesn't need touching when the attribute lands. I'd do it as the next PR, spelling to be bikeshed (`@testable`?), gated experimental like `@shadowable` to start.

On writing: the sketch is a third vtable entry mirroring the read side. It parses back the same string encoding `element_property_value` emits, same tri-state, same debug-info gate. Generated code dispatches through the same table to a compiled setter; the interpreter goes through `store_property`; MCP gains a `set_element_property` tool. The one semantic question is binding replacement: a set from the test channel would behave like a set from the language APIs (replace the binding with a value), which is what UI tests generally want. I kept it out of this PR so the read contract stays reviewable on its own.

**Comment by tronical** [comment]
The essence:

>  I'd do it as the next PR, spelling to be bikeshed (@testable?), gated experimental like @shadowable to start.

I can't believe I'm saying this, but perhaps this would go well in a single PR :-)

**Comment by tilladam** [comment]
Done: the attribute is in this PR now, as four follow-up commits on top of the reviewed head (front end, retention/pinning, tests, docs), so the read channel and the attribute stay reviewable as two layers. Spelling is `@testable`, experimental-gated like `@shadowable`; renaming is cheap until merge if you prefer something else.

What it guarantees, per your "not necessarily public API, but relevant to testing":

- listed in the declared-property table regardless of visibility, `private` included, with language visibility unchanged;
- survives both removal stages (object-tree `remove_unused_properties` and the LLR `remove_unused`), so a future optimization can't silently drop it;
- listing and a live binding, not a readable value for every type — a `@testable` struct lists with no value like any other struct;
- only when debug info is on. Without it, the property is removed like any other unused one, so release builds don't change (pinned by a unit test).

Only properties accept it; callbacks, functions, global and interface properties get a diagnostic. Docs under `guide/experimental/testable`. The PR description has the details.

File: docs/development/mcp-server.md [hunk 1]
```diff
@@ -47,6 +47,22 @@ The lazy start via `OnceCell` ensures the server only binds the port once the ap
...
+(see `SubComponent::element_properties` in the compiler's LLR).
+Values cross the vtable as strings in the `i_slint_core::debug_info` encoding;
+`mcp_server.rs::convert_declared_property_values` retypes them into JSON booleans and
+numbers for both `get_element_properties` and `get_element_tree`.
+The table exists only when the application was compiled with debug info, and an item tree
+whose generator does not implement the entries (generated C++ today) reports "unsupported",
+which surfaces as a `note` instead of an empty list.
+A property normally needs `in`/`out`/`in-out` visibility and a live binding to appear here;
+the experimental `@testable` attribute (`docs/astro/.../guide/experimental/testable.mdx`)
+overrides both, so a `private` or otherwise-unread property can still be listed and read.
```
Comment (tronical) [inline docs/development/mcp-server.md:64]: I think this is where I might not have explained it well enough. The idea was that only @testable properties are exposed. A regular in-out property for example could still be optimized away and disappear - if unused, while I think we'd want the output in the MCP to be reliable and stable.

Comment (tilladam) [reply docs/development/mcp-server.md:64]: Got it: the table is exactly the `@testable` set, nothing else. That makes the listing independent of the optimizer, and the "may be optimized away" caveat goes away with it. I'll invert the filter and rework the tests.

Two things to settle while I'm at it:

1. With only annotated properties in it, `declared_properties` / `declaredProperties` is the wrong name. I'd rename to `testable_properties` / `testableProperties` across Rust, C++ and the MCP response now, while it's cheap.
2. `@testable` is currently experimental-gated like `@shadowable`. Since it becomes the only way into the channel, that gates the whole MCP property read behind `SLINT_ENABLE_EXPERIMENTAL_FEATURES`. Do you still want the gate?

Comment (tronical) [reply docs/development/mcp-server.md]: > Got it: the table is exactly the `@testable` set, nothing else. That makes the listing independent of the optimizer, and the "may be optimized away" caveat goes away with it. I'll invert the filter and rework the tests.
> 
> Two things to settle while I'm at it:
> 
> 1. With only annotated properties in it, `declared_properties` / `declaredProperties` is the wrong name. I'd rename to `testable_properties` / `testableProperties` across Rust, C++ and the MCP response now, while it's cheap.

Yes, exactly.

> 2. `@testable` is currently experimental-gated like `@shadowable`. Since it becomes the only way into the channel, that gates the whole MCP property read behind `SLINT_ENABLE_EXPERIMENTAL_FEATURES`. Do you still want the gate?

For the time being I think that makes sense. Makes it easier to get to the point where we could land this in master :)

Comment (tilladam) [reply docs/development/mcp-server.md]: Done in the latest commit: the table is exactly the `@testable` set, and everything is named after it now (`testable_properties()` / `testable_property_value()` in Rust and C++, `testableProperties` in the MCP response, `TestableProperty` in the proto). The experimental gate stays. The shared test case annotates what it observes and checks that an unannotated public property is not listed; the `changed` callbacks that kept properties alive are gone, since the attribute does that now. PR description updated.
